### PR TITLE
[New Scheduler] Implement InvokerHealthyManager

### PR DESCRIPTION
**Pull request by ningyougang**
## Description
Manage New scheduler's invoker healthy management
Design document: https://cwiki.apache.org/confluence/display/OPENWHISK/InvokerHealthyManager

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by bdoyle0182** [comment]
Now that we are beginning to touch the existing components, we should send out a message on the mailing list that new changes are entering the existing code that are not yet in use as the scheduler is slowly being rolled out in code.

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (rabbah) [inline common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:446]: is this method needed? cases class equals already provides this implementation by definition.

File: common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/core/service/DataManagementService.scala:5]: I hope this PR is rebased after this kind of module is landed according to this plan:
https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala]: @ningyougang 
IIRC, we introduced this not to store data to ETCD if the data is the same as the previous one.
Can we make sure this requirement with the default comparison functionality provided by the case class?

I am not quite sure but we were using this class without these methods but introduced them at some point for some reason.

Could you confirm once?

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala]: Ah, yes, i tested in my local, after removed `the override def equals and the override def hashCode`, works well both, e.g.
for two InvokerResourceMessage object,
* If all field value is same, object1.equals(object) will be `true`, 
* if any one filed value is different,  object1.equals(object) will be `false`

File: common/scala/src/main/scala/org/apache/openwhisk/core/entity/InstanceId.scala
Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/entity/InstanceId.scala:40]: Good question.

I tested in my local, doesn't affect the message bus.  This pr's invoker in upsteam master's controller's healthy status is `up`.
But your said problem i meet before, in that time, it seems the PingMessage is changed, so lead to the invoker of new codes in controller of old code's healthy status is `unhealthy`. if this issue comes. we solved it using another deployment method, e.g.
* Remove half controller from nginx.
* Disable half invoker
* Deploy half controller/invoker using new codes
* Add the half controller nginx.
* Deploy another half components using above steps

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala]: I already removed the equals and hashCode.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala]: Updated accordingly

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala]: Updated accordingly

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala]: Ok, i will check

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:435]: This is used for some special actions. e.g. need high memory, need power gpu.
We can make the action's invocations run on corresponding invokers

* Firstly, add some annotation to the actions, e.g.

```
wsk action create hello-gpu ~/hello-gpu.js --annotation  invoker-resources ["gpu"]
```
* Secondly, when deploy invoker, add the relative tag to the invoker as well
```
whisk/invokers/0/0
{"busyMemory":0,"dedicatedNamespaces":[],"freeMemory":10240,"inProgressMemory":0,"status":"up","tags":['gpu']}
whisk/invokers/1/1
{"busyMemory":0,"dedicatedNamespaces":[],"freeMemory":10240,"inProgressMemory":0,"status":"up","tags":[]}
```
* Finally, when run the action, the activations for that action will run on above invoker0 which includes tag: gpu

Comment (ningyougang) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:435]: BTW, dedicatedNamespaces means dedicatedNamespaces's all actions ran on corresponding invoker.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala:54]:
```suggestion
case object CreatingContainer extends ProxyState
```

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala:56]:
```suggestion
case object CreatingClient extends ProxyState
```

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:56]: nit: There's a lot of unnecessary blank lines from here on down we can clean up

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:218]: In the existing invoker supervision, the buffer includes user actions result for the ring buffer to check for system error. I don't know without seeing more of the new scheduler code, but we'll still have those checks right elsewhere? Also since this ring buffer seems to only include health check activations, can this be more aggressive? I would think you can just go unhealthy if a single one fails rather than three of the last ten.

File: common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala
Comment (bdoyle0182) [reply common/scala/src/main/scala/org/apache/openwhisk/core/connector/Message.scala:435]: This is a VERY powerful feature to group invokers depending on the operator's needs without creating multiple clusters. I'm glad it's being introduced with the new scheduler.

**Review by bdoyle0182 — COMMENTED** [review]
I think this generally looks good to me. But I think I need to see the the upstream DataManagementService this is forwarding message to and possibly the new downstream ContainerProxy that I believe is requesting activations by sending a message to this fsm?

So is my understanding correct that every invoker will have a healthcheck container running at all times? What is the gap between requesting healthcheck activations from this fsm from the container or is it just always trying to run the healthcheck code with no gaps?

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:56]: cleaned up

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:209]: Currently, i used a simple method to check invoker's health status.
did't check for system error.

Maybe after this pr merged, we can submit another pr to optimize it. (I din't have much more good idea for it.)

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala]: I removed `var useMemory = 0l`

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala]: Updated accordingly.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/FunctionPullingContainerProxy.scala]: Updated accordingly

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:56]: Already cleaned up.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:209]: Currently, i used a simple method to check invoker's health status.
did't check for system error.

**Comment by ningyougang** [comment]
@bdoyle0182 

> I think this generally looks good to me. But I think I need to see the the upstream DataManagementService this is forwarding message to and possibly the new downstream ContainerProxy that I believe is requesting activations by sending a message to this fsm?

Sorry, i don't understand what you mean. i tried to share like below
This is the DataManagementService pr: https://github.com/apache/openwhisk/pull/5063, store data to etcd with lease / delete data from etcd.  invokerHealthyManager will use DataManagementService  to store its healthy data to etcd,  e.g. dataManagementService ! UpdateDataOnChange(InvokerKeys.health(instanceId), invokerResourceMessage.serialize)


> So is my understanding correct that every invoker will have a healthcheck container running at all times? What is the gap between requesting healthcheck activations from this fsm from the container or is it just always trying to run the healthcheck code with no gaps?

InvokerHealthyManager doesn't have a healthcheck container running at all time, it will create a healthcheck container at the very beginning or when the healthy status's changed from Healthy to Unhealthy, for the activation request for the healthcheck container,  it will fetch the local activation, e.g.
image
BTW, for the other action container, it will fetch activation from scheduler component's memoryQueue using grpc.

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (bdoyle0182) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:141]: nit: can't this be removed? you can just leave out state transitions that don't do anything for this partial function

**Comment by bdoyle0182** [comment]
LGTM but this should wait until rebasing with DataManagementService to merge

**Review by bdoyle0182 — APPROVED** [review]

**Review by bdoyle0182 — APPROVED** [review]

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (style95) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:141]: +1

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (style95) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:209]: We just preserved the existing health check logic but I think we can make it configurable.
Each container proxy is supposed to send a message indicating a failed result to InvokerHealthManager

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (style95) [inline core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala:1]: I think the filename should be `InvokerHealthManager`?

File: tests/src/test/scala/org/apache/openwhisk/core/containerpool/v2/test/InvokerHealthyManagerTests.scala
Comment (style95) [inline tests/src/test/scala/org/apache/openwhisk/core/containerpool/v2/test/InvokerHealthyManagerTests.scala:1]: The filename should be `InvokerHealthManagerTests`?

**Review by style95 — COMMENTED** [review]
@ningyougang Could you rebase this PR once as this PR gets merged?
https://github.com/apache/openwhisk/pull/5063

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala]: Already fixed

File: core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala
Comment (ningyougang) [reply core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/v2/InvokerHealthyManager.scala]: Already changed the file name

File: tests/src/test/scala/org/apache/openwhisk/core/containerpool/v2/test/InvokerHealthyManagerTests.scala
Comment (ningyougang) [reply tests/src/test/scala/org/apache/openwhisk/core/containerpool/v2/test/InvokerHealthyManagerTests.scala]: Already changed the file name

**Comment by ningyougang** [comment]
Already rebased the latest code, and ready to review again.

**Review by style95 — APPROVED** [review]

**Comment by ningyougang** [comment]
have any commit, if this pr merged, i can submit subsequent feature

**Comment by style95** [comment]
I suppose Brendan already approved this so we can merge this and move forward?

**Pull request merged**
